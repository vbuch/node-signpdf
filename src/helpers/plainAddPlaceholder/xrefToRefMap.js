import SignPdfError from '../../SignPdfError';

const xrefToRefMap = (xrefString) => {
    const lines = xrefString.split('\n').filter((l) => l !== '');

    let index = 0;
    let expectedLines = 0;
    const xref = new Map();
    lines.forEach((line) => {
        const split = line.split(' ');
        if (split.length === 2) {
            index = parseInt(split[0]);
            expectedLines = parseInt(split[1]);
            return;
        }
        if (expectedLines <= 0) {
            throw new SignPdfError(
                'Too many lines in xref table.',
                SignPdfError.TYPE_PARSE,
            );
        }
        expectedLines -= 1;
        const [offset, , inUse] = split;
        if (inUse.trim() === 'f') {
            index += 1;
            return;
        }
        if (inUse.trim() !== 'n') {
            throw new SignPdfError(
                `Unknown in-use flag "${inUse}". Expected "n" or "f".`,
                SignPdfError.TYPE_PARSE,
            );
        }
        if (!/^\d+$/.test(offset.trim())) {
            throw new SignPdfError(
                `Expected integer offset. Got "${offset}".`,
                SignPdfError.TYPE_PARSE,
            );
        }
        const storeOffset = parseInt(offset.trim());
        xref.set(index, storeOffset);
        index += 1;
    });

    return xref;
};

export default xrefToRefMap;
