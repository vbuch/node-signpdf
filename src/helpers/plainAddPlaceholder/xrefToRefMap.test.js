import SignPdfError from '../../SignPdfError';
import xrefToRefMap from './xrefToRefMap';

const xrefStrings = [
    `0 39
0000000000 65535 f 
0000100423 00000 n 
0000100501 00000 n 
0000100361 00000 n 
0000100340 00000 n 
0000061908 00000 n 
0000061783 00000 n 
0000061679 00000 n 
0000099156 00000 n 
0000121651 00000 n 
0000000015 00000 n 
0000062401 00000 n 
0000062300 00000 n 
0000062193 00000 n 
0000097854 00000 n 
0000064460 00000 n 
0000064370 00000 n 
0000064263 00000 n 
0000097049 00000 n 
0000096948 00000 n 
0000096824 00000 n 
0000096650 00000 n 
0000096706 00000 n 
0000100195 00000 n 
0000097749 00000 n 
0000097600 00000 n 
0000097625 00000 n 
0000097650 00000 n 
0000097686 00000 n 
0000097718 00000 n 
0000100548 00000 n 
0000097952 00000 n 
0000098210 00000 n 
0000098776 00000 n 
0000113292 00000 n 
0000099303 00000 n 
0000099567 00000 n 
0000099874 00000 n 
0000115410 00000 n `,
    `0 1
0000000000 65535 f
6 2
0000662716 00000 n
0000662578 00000 n
39 35
0000655893 00000 n
0000656101 00000 n
0000656080 00000 n
0000656127 00000 n
0000656181 00000 n
0000656389 00000 n
0000656368 00000 n
0000656415 00000 n
0000656469 00000 n
0000656794 00000 n
0000656771 00000 n
0000672037 00000 n
0000656877 00000 n
0000657136 00000 n
0000657114 00000 n
0000661923 00000 n
0000657252 00000 n
0000659653 00000 n
0000659629 00000 n
0000661899 00000 n
0000662162 00000 n
0000662140 00000 n
0000662250 00000 n
0000662498 00000 n
0000662476 00000 n
0000663052 00000 n
0000662932 00000 n
0000663030 00000 n
0000663185 00000 n
0000663207 00000 n
0000672013 00000 n
0000672535 00000 n
0000672943 00000 n
0000672920 00000 n
0000673269 00000 n`,
];

describe('xrefToRefMap', () => {
    it('Predefined xrefs match their snapshots', () => {
        xrefStrings.forEach((xref) => {
            expect(xrefToRefMap(xref)).toMatchSnapshot();
        });
    });
    it('Throws an error when number of items does not match statement', () => {
        try {
            xrefToRefMap(`0 1
000 000 f
001 000 n
002 100 n`);
            expect('here').not.toBe('here');
        } catch (e) {
            expect(e instanceof SignPdfError).toBe(true);
            expect(e.type).toBe(SignPdfError.TYPE_PARSE);
        }
    });
    it('Throws an error when unknown in-use flag is uses', () => {
        try {
            xrefToRefMap(`0 3
000 000 f
001 000 n
002 100 w`);
            expect('here').not.toBe('here');
        } catch (e) {
            expect(e instanceof SignPdfError).toBe(true);
            expect(e.type).toBe(SignPdfError.TYPE_PARSE);
            expect(e.message).toMatchSnapshot();
        }
    });
    it('Throws an error when non-integer offset is given', () => {
        try {
            xrefToRefMap(`0 3
000 000 f
001 000 n
20a 100 n`);
            expect('here').not.toBe('here');
        } catch (e) {
            expect(e instanceof SignPdfError).toBe(true);
            expect(e.type).toBe(SignPdfError.TYPE_PARSE);
            expect(e.message).toMatchSnapshot();
        }
    });
});
