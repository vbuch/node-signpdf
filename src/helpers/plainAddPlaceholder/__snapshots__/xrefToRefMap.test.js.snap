// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xrefToRefMap Predefined xrefs match their snapshots 1`] = `
Map {
  1 => 100423,
  2 => 100501,
  3 => 100361,
  4 => 100340,
  5 => 61908,
  6 => 61783,
  7 => 61679,
  8 => 99156,
  9 => 121651,
  10 => 15,
  11 => 62401,
  12 => 62300,
  13 => 62193,
  14 => 97854,
  15 => 64460,
  16 => 64370,
  17 => 64263,
  18 => 97049,
  19 => 96948,
  20 => 96824,
  21 => 96650,
  22 => 96706,
  23 => 100195,
  24 => 97749,
  25 => 97600,
  26 => 97625,
  27 => 97650,
  28 => 97686,
  29 => 97718,
  30 => 100548,
  31 => 97952,
  32 => 98210,
  33 => 98776,
  34 => 113292,
  35 => 99303,
  36 => 99567,
  37 => 99874,
  38 => 115410,
}
`;

exports[`xrefToRefMap Predefined xrefs match their snapshots 2`] = `
Map {
  6 => 662716,
  7 => 662578,
  39 => 655893,
  40 => 656101,
  41 => 656080,
  42 => 656127,
  43 => 656181,
  44 => 656389,
  45 => 656368,
  46 => 656415,
  47 => 656469,
  48 => 656794,
  49 => 656771,
  50 => 672037,
  51 => 656877,
  52 => 657136,
  53 => 657114,
  54 => 661923,
  55 => 657252,
  56 => 659653,
  57 => 659629,
  58 => 661899,
  59 => 662162,
  60 => 662140,
  61 => 662250,
  62 => 662498,
  63 => 662476,
  64 => 663052,
  65 => 662932,
  66 => 663030,
  67 => 663185,
  68 => 663207,
  69 => 672013,
  70 => 672535,
  71 => 672943,
  72 => 672920,
  73 => 673269,
}
`;

exports[`xrefToRefMap Throws an error when non-integer offset is given 1`] = `"Expected integer offset. Got \\"20a\\"."`;

exports[`xrefToRefMap Throws an error when unknown in-use flag is uses 1`] = `"Unknown in-use flag \\"w\\". Expected \\"n\\" or \\"f\\"."`;
